android/vcmi-app/src/main/java/eu/vcmi/vcmi/NativeMethods.java

```diff
@@ -104,4 +111,29 @@ private static Messenger requireServerMessenger()
         }
         return msg;
     }
+
+    @SuppressWarnings(Const.JNI_METHOD_SUPPRESS)
+    public static void copyFileFromUri(String sourceFileUri, String destinationFile)
```

Comment:
it should go into `util/FileUtil.java`. And that file already contains basically the same code inside the directory copy method, just extract it.